src/GenSymIO.chpl

```diff
@@ -730,30 +797,319 @@ module GenSymIO {
+
+        file_id = C_HDF5.H5Fcreate(filenames[loc].c_str(), C_HDF5.H5F_ACC_TRUNC, 
+                                                         C_HDF5.H5P_DEFAULT, C_HDF5.H5P_DEFAULT);
+
         if file_id < 0 { // Negative file_id means error
           throw new owned FileNotFoundError();
+        }
+          /*
+           * If DType is UInt8, need to create strings_array group to enable read/load with the
+           * Arkouda infrastructure. The strings_array group contains two datasets: (1) segments, 
+           * which are the indices for the string values embedded in the string binary and (2)
+           * values, which are the corresponding string values within a null-delimited bytes object
+           */ 
+           if array_type == DType.UInt8 {
+             var group_id = C_HDF5.H5Gcreate2(file_id, "/strings_array", C_HDF5.H5P_DEFAULT, 
+                                              C_HDF5.H5P_DEFAULT, C_HDF5.H5P_DEFAULT);
+             C_HDF5.H5Gclose(group_id);
         }
         C_HDF5.H5Fclose(file_id);
       }
     }
-    coforall (loc, idx) in zip(A.targetLocales(), filenames.domain) do on loc {
+
+    /*
+     * Declare the indices object, which is a globally-scoped PrivateSpace array that
+     * contains the slice index for each locale. Each slice index is used to remove 
+     * the uint(8) characters moved to the previous locale, which is done when a 
+     * string, which is an array of uint(8) chars, spans two locales.
+     */
+    var indices: [PrivateSpace] int;
+    
+    /*
+     * If this is a strings dataset, loop through all locales and set the slice indices, 
+     * which are used to remove uint(8) characters from the locale slice that are part 
+     * of a string that belongs to the previous locale in the pdarray list of locales.
+     */
+    if isStringsDataset(dsetName) {
+      coforall (loc, idx) in zip(A.targetLocales(), 
+                                       filenames.domain) with(ref indices) do on loc {
+      if idx < A.targetLocales().size-1 {
+          const locDom = A.localSubdomain();
+          if A.localSlice(locDom).back() != NULL_STRINGS_VALUE {
+            generateSliceIndex(idx,indices,A);
+          }
+        }
+      }
+    }
+
+    coforall (loc, idx) in zip(A.targetLocales(), 
+                                        filenames.domain) with(ref indices) do on loc {
         const myFilename = filenames[idx];
         if GenSymIO_DEBUG {
           writeln(try! "%s exists? %t".format(myFilename, exists(myFilename)));
         }
-        var myFileID = C_HDF5.H5Fopen(myFilename.c_str(), C_HDF5.H5F_ACC_RDWR, C_HDF5.H5P_DEFAULT);
+        var myFileID = C_HDF5.H5Fopen(myFilename.c_str(), C_HDF5.H5F_ACC_RDWR, 
+                                                                    C_HDF5.H5P_DEFAULT);
         const locDom = A.localSubdomain();
         var dims: [0..#1] C_HDF5.hsize_t;
         dims[0] = locDom.size: C_HDF5.hsize_t;
```

Comment:
if `segments` arg is set, then we are writing Strings bytes and need to cut based on locality of `segments`. Here's roughly how I would do it:

```chapel
const mySegDom = segments.localSubdomain();
var myEnd: int;
if mySegDom.high < segments.domain.high {
  // If not the last locale, go to the start of the next locale's first segment, then backtrack one
  myEnd = segments[mySegDom.high+1] - 1;
} else {
  // Last locale goes to the end of bytes array
  myEnd = A.domain.high;
}
const locDom = {mySegDom.low..myEnd};
// This copies everything in locDom to a local buffer, which is required for c_ptrTo() later
const locSlice: [locDom] A.eltType = A[locDom];
var dims: [0..#1] C_HDF5.hsize_t;
dims[0] = locDom.size: C_HDF5.hsize_t;
H5LTmake_dataset_WAR(myFileID, myDsetName.c_str(), 1, c_ptrTo(dims), getHDF5Type(A.eltType), c_ptrTo(locSlice));
```

When `tohdfMsg` gets called on a Strings array, it should call `write1DDistArray` twice (see above comment): the first time on the offsets/segments array, and the second time on the bytes/values array. This code should only activate on that second call. However, that means that the first call somehow needs to know to create the group. I haven't worked through how to do that in my comments.

Reply:
I have the logic to create the group, no worries. 

So there's a larger issue with segments, something that I discussed with @bradcray, @e-kayrakli, and Tom Rollinger. The key challenge as we've discussed is that a string, which, as we know in Arkouda is actually an array of non-null uint(8) characters, can span two locales, whereas other datatypes such as floats and ints don't; there's basically two ways to deal with this: (1) addressing how strings are partitioned within locales (which I'd prefer) or (2) reassembling the strings AKA uint(8) sequences and updating or assignment segments prior to writing out to hdf5.

I initially investigated option 1. The way I understand it, the boundingBox is designed to divide an array up evenly over all locales. I asked in the Chapel gitter channel if there's a way to configure Chapel to partition things differently so string elements, which are simply non-null uint(8) characters in the case of Arkouda, end up on the same locale. In other words, is there a way to ensure strings are not spread across two adjacent locales. It's my understanding from Brad's and  Engin's comments that this would be a large undertaking, definitely way beyond what I know about Chapel, so I turned my attention to option 2.

To deliver on option 2, there's two situations that need to be accounted for: (1) a values array within a locale ending with a non-null uint(8) char, which, in my testing, indicates that part of a string is in the current locale and the rest is in the next locale and (2) a values array in a locale that contains 1..n non-null uint(8) chars corresponding to a string started in the previous locale. In my testing, virtually every values array in every locale is affected by some combination of situations 1 and 2, because an occurrence of situation 1 acts as a domino effect. There are two direct consequences of this. Firstly, the segments for each locale can't be known until situations (1) and (2) related to partitioning of strings across locales are resolved. Secondly and consequently, whereas originally I was sending the offsets (as named in the Python code) AKA segments and persisting those, now I just send the values and generate the segments once the values are correctly partitioned so that each string--series of non-null uint(8) characters--are written to a single hdf5 file mapping to a single locale. Again, let's whiteboard this.